File: benchmark/c/cpw/write_an_efficient_method_to_check_if_a_number_is_multiple_of_3/write_an_efficient_method_to_check_if_a_number_is_multiple_of_3.c


#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>



int write_an_efficient_method_to_check_if_a_number_is_multiple_of_3 ( int n ) ;
int write_an_efficient_method_to_check_if_a_number_is_multiple_of_3 ( int n ) {
  int odd_count = 0;
  int even_count = 0;
  if ( n < 0 ) n = - n;
  if ( n == 0 ) return 1;
  if ( n == 1 ) return 0;
  while ( n ) {
    if ( n & 1 ) odd_count ++;
    if ( n & 2 ) even_count ++;
    n = n >> 2;
  }
  return write_an_efficient_method_to_check_if_a_number_is_multiple_of_3 ( abs ( odd_count - even_count ) );
}

File: translation_result/else_bms/fail/c/cpw/write_an_efficient_method_to_check_if_a_number_is_multiple_of_3/gemini/Hinted/write_an_efficient_method_to_check_if_a_number_is_multiple_of_3.rs

fn write_an_efficient_method_to_check_if_a_number_is_multiple_of_3(n: i32) -> bool {
    let mut odd_count = 0;
    let mut even_count = 0;
    let mut n = n;
    if n < 0 {
        return write_an_efficient_method_to_check_if_a_number_is_multiple_of_3(-n);
    }
    if n == 0 {
        return true;
    }
    if n == 1 {
        return false;
    }
    while n > 0 {
        if n % 2 == 1 {
            odd_count += 1;
        }
        if n % 4 == 2 {
            even_count += 1;
        }
        n >>= 2;
    }
    let diff: i32 = odd_count - even_count;
    write_an_efficient_method_to_check_if_a_number_is_multiple_of_3(diff.abs() as i32 % 3)
}
